### PR TITLE
recursive-patterns(10): Rename elements of the pattern-matching implementation for clarity

### PR DESCRIPTION
**Pull request by gafter**
Also add documentation.

Only the third-to-last iteration is unique to this PR. The second-to-last iteration is an integration from upstream, and the last is a tiny doc bug fix.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs [hunk 1]
```diff
@@ -12,8 +12,46 @@
...
+    /// temporary variables for use in subsequent nodes (think of it as a node with a single successor);
+    /// <see cref="BoundWhenDecisionDagNode"/> represents the test performed by evaluating the expression of the
+    /// when-clause of a switch case; and <see cref="BoundLeafDecisionDagNode"/> represents a leaf node when we
+    /// have finally determined exactly which case matches. Each test processes a single input, and there are
+    /// four kinds:<see cref="BoundDagNullTest"/> tests a value for null; <see cref="BoundDagNonNullTest"/>
+    /// tests that a value is not null; <see cref="BoundDagTypeTest"/> checks if the value is of a given type;
+    /// and <see cref="BoundDagValueTest"/> checks if the value is equal to a given constant. Of the evaluations,
+    /// there are <see cref="BoundDagDeconstructEvaluation"/> which represents an invocation of a type's
+    /// "Deconstruct" method; <see cref="BoundDagFieldEvaluation"/> reads a field; <see cref="BoundDagPropertyEvaluation"/>
+    /// reads a property; and <see cref="BoundDagTypeEvaluation"/> converts a value from one type to another (which
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs]: >BoundDagTypeEvaluation [](start = 41, length = 22)

Would it make sense to call this `BoundDagCastEvaluation`?

Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs]: Well, it only converts a value to a type that it already has at runtime. A cast does so much more.

**Comment by gafter** [comment]
@agocke I think this one should be easy to review. It is largely based on your suggestions.

File: src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs [hunk 2]
```diff
@@ -671,23 +691,23 @@ BoundDecisionDag finalState(SyntaxNode syntax, LabelSymbol label, ImmutableArray
         }
 
         private void FilterCase(
-            PartialCaseDecision c,
-            BoundDagDecision d,
-            ArrayBuilder<PartialCaseDecision> whenTrueBuilder,
-            ArrayBuilder<PartialCaseDecision> whenFalseBuilder)
+            RemainingTestsForCase c,
+            BoundDagTest d,
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs]: "d" => "test"

File: src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs [hunk 3]
```diff
@@ -716,16 +736,16 @@ BoundDecisionDag finalState(SyntaxNode syntax, LabelSymbol label, ImmutableArray
                 }
             }
 
-            PartialCaseDecision makeNext(ArrayBuilder<BoundDagDecision> remainingDecisions)
+            RemainingTestsForCase makeNext(ArrayBuilder<BoundDagTest> remainingDecisions)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs]: "remainingDecisions" -> "remainingTests"

File: src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs [hunk 4]
```diff
@@ -51,28 +52,36 @@ public HashSet<LabelSymbol> ReachableLabels
...
+                        this.TopologicallySortedNodes.OfType<BoundLeafDecisionDagNode>().Select(d => d.Label));
                 }
 
                 return _reachableLabels;
             }
         }
 
         /// <summary>
-        /// Return a list of all the nodes reachable from this one, in a topologically sorted order.
+        /// A list of all the nodes reachable from this one, in a topologically sorted order.
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs]: Should say "reachable from the root"

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs [hunk 5]
```diff
@@ -41,20 +41,20 @@ public static BoundNode Rewrite(LocalRewriter localRewriter, BoundPatternSwitchS
             private Dictionary<SyntaxNode, LabelSymbol> _sectionLabels = PooledDictionary<SyntaxNode, LabelSymbol>.GetInstance();
 
             /// <summary>
-            /// We revise the returned label for a decision so that all decisions in the same switch section are given the same label.
-            /// This permits all the decisions in a
+            /// We revise the returned label for a leaf so that all leafs in the same switch section are given the same label.
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_PatternSwitchStatement.cs]: leafs -> leaves

**Pull request merged**